src/uct/ib/dc/dc_mlx5.c

```diff
@@ -808,45 +809,78 @@ void uct_dc_mlx5_destroy_dct(uct_dc_mlx5_iface_t *iface)
+{
+    ucs_arbiter_cleanup(&dci_pool->arbiter);
+    ucs_free(dci_pool->stack);
+}
+
+static void uct_dc_mlx5_iface_dcis_destroy(uct_dc_mlx5_iface_t *iface,
+                                           uint8_t num_dci_pools,
+                                           uint8_t num_dcis)
 {
     uct_ib_mlx5_md_t *md = ucs_derived_of(iface->super.super.super.super.md,
                                           uct_ib_mlx5_md_t);
-    int i;
+    uint8_t pool_index, dci_index;
+
+    ucs_assert(num_dci_pools <= iface->tx.num_dci_pools);
+    ucs_assert(num_dcis <= uct_dc_mlx5_iface_total_ndci(iface));
 
-    for (i = 0; i < max; i++) {
-        uct_rc_txqp_cleanup(&iface->super.super, &iface->tx.dcis[i].txqp);
-        uct_ib_mlx5_destroy_qp(md, &iface->tx.dcis[i].txwq.super);
+    for (dci_index = 0; dci_index < num_dcis; dci_index++) {
+        uct_rc_txqp_cleanup(&iface->super.super,
+                            &iface->tx.dcis[dci_index].txqp);
+        uct_ib_mlx5_destroy_qp(md, &iface->tx.dcis[dci_index].txwq.super);
+    }
+
+    for (pool_index = 0; pool_index < num_dci_pools; pool_index++) {
+        uct_dc_mlx5_iface_dci_pool_destroy(&iface->tx.dci_pool[pool_index]);
     }
 
     ucs_free(iface->tx.dcis);
 }
 
-ucs_status_t
-uct_dc_mlx5_iface_create_dcis(uct_dc_mlx5_iface_t *iface,
+static ucs_status_t
+uct_dc_mlx5_iface_dcis_create(uct_dc_mlx5_iface_t *iface,
                               const uct_dc_mlx5_iface_config_t *config)
 {
-    uint8_t num_paths = iface->super.super.super.num_paths;
+    const uint8_t num_paths = iface->super.super.super.num_paths;
+    uint8_t dci_index       = 0;
     uct_dc_mlx5_dci_pool_t *dci_pool;
-    int i, pool_index, dci_index;
+    uint8_t pool_index, i;
     ucs_status_t status;
 
-    dci_index = 0;
+    iface->tx.dcis = ucs_calloc((iface->tx.ndci * iface->tx.num_dci_pools) +
+                                UCT_DC_MLX5_KEEPALIVE_NUM_DCIS,
+                                sizeof(*iface->tx.dcis),
+                                "dcis");
+    if (iface->tx.dcis == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err;
+    }
+
     for (pool_index = 0; pool_index < iface->tx.num_dci_pools; pool_index++) {
-        dci_pool = &iface->tx.dci_pool[pool_index];
-        ucs_debug("creating dci pool %d with %d QPs", pool_index,
-                  iface->tx.ndci);
+        ucs_debug("creating dci pool %u with %u QPs", pool_index, iface->tx.ndci);
+
+        iface->tx.dci_pool[pool_index].stack = ucs_calloc(iface->tx.ndci,
+                                                          sizeof(uint8_t),
+                                                          "dci pool stack");
+        if (iface->tx.dci_pool[pool_index].stack == NULL) {
+            status = UCS_ERR_NO_MEMORY;
+            goto err_dcis_destroy;
+        }
+
+        ucs_arbiter_init(&iface->tx.dci_pool[pool_index].arbiter);
+        dci_pool                    = &iface->tx.dci_pool[pool_index];
```

Comment:
why not initialize dci_pool earlier like was done before this PR?

Reply:
to have it in a single function and then use a single function `uct_dc_mlx5_iface_dcis_destroy` to destroy them all

Reply:
still, we could use dci_pool instead of &iface->tx.dci_pool[pool_index] in few places